llvm/lib/Analysis/UniformityAnalysis.cpp

```diff
@@ -152,26 +152,34 @@ UniformityInfoWrapperPass::UniformityInfoWrapperPass() : FunctionPass(ID) {
 INITIALIZE_PASS_BEGIN(UniformityInfoWrapperPass, "uniformity",
                       "Uniformity Analysis", true, true)
 INITIALIZE_PASS_DEPENDENCY(DominatorTreeWrapperPass)
-INITIALIZE_PASS_DEPENDENCY(CycleInfoWrapperPass)
 INITIALIZE_PASS_DEPENDENCY(TargetTransformInfoWrapperPass)
 INITIALIZE_PASS_END(UniformityInfoWrapperPass, "uniformity",
                     "Uniformity Analysis", true, true)
 
 void UniformityInfoWrapperPass::getAnalysisUsage(AnalysisUsage &AU) const {
   AU.setPreservesAll();
   AU.addRequired<DominatorTreeWrapperPass>();
-  AU.addRequiredTransitive<CycleInfoWrapperPass>();
   AU.addRequired<TargetTransformInfoWrapperPass>();
 }
 
 bool UniformityInfoWrapperPass::runOnFunction(Function &F) {
-  auto &cycleInfo = getAnalysis<CycleInfoWrapperPass>().getResult();
   auto &domTree = getAnalysis<DominatorTreeWrapperPass>().getDomTree();
   auto &targetTransformInfo =
       getAnalysis<TargetTransformInfoWrapperPass>().getTTI(F);
 
   m_function = &F;
-  m_uniformityInfo = UniformityInfo{domTree, cycleInfo, &targetTransformInfo};
+
+  // FIXME: CycleInfo is a transitive dependency for UniformityInfo, and we
+  // would have liked to say that CycleInfo is also preserved whenever
+  // UniformityInfo is preserved. But some passes like
+  // AMDGPUUnifyDivergentExitNodes do not update cycle membership and lists of
+  // cycle exits, since that facility is not available. The end result is that
+  // we cannot declare CycleInfo as a transitive depedency for UniformityInfo.
+  // Instead we recompute CycleInfo on demand. This is OK because it is rarely
+  // needed, and only affects the old pass manager.
+  m_cycleInfo.clear();
+  m_cycleInfo.compute(F);
+  m_uniformityInfo = UniformityInfo{domTree, m_cycleInfo, &targetTransformInfo};
```

Comment:
The transitive dependency means that CycleInfo must be present and correct not just when UniformityInfo is run, but whenever it is queried. If UnifyDivergentExitNodes claims to preserve UniformityInfo but does not update CycleInfo then it sounds like there is a real problem there: as soon as UnifyDivergentExitNodes has made some changes, a query on UniformityInfo will no longer be reliable because CycleInfo will not be up to date. I don't see how computing CycleInfo in this "run" function fixes that problem.

Reply:
I always find transitive dependencies in the old pass manager confusing. It is supposed to mean what you said, but the actual implementation is not clear to me.

Specifically for UniformityInfo, CycleInfo is used only during its computation and does not need to be present during the lifetime of the UniformityInfo itself. But then there is this implicit assumption in general, if UniformityInfo is updated, then CycleInfo may need to be present as a dependency. This is not true in this case because UnifyDivergentExitNodes doesn't actually update UniformityInfo in any way. If all this is implemented correctly, then CycleInfo does not need to be marked transitive ... it is not needed to be alive during the lifetime of UniformityInfo.

Actually @Pierre-vh, you had marked CycleInfo as a transitive dependency at some point; it must have fixed something for him. Do you remember what that was?

Reply:
Oh wait, the query `isTemporalDivergent()` needs CycleInfo, during the lifetime of UniformityInfo. I'll need to rethink my approach here ... either claim that a local copy of CycleInfo is sufficient, or preserve UniformityInfo only when it makes sense to also preserve CycleInfo.

Reply:
I'd expect the same sort of CFG transforms to invalidate both, so intuitively to me it makes sense they would always be paired